packages/decentraland-ecs/src/decentraland/UIShapes.ts

```diff
@@ -350,17 +350,36 @@ export class UIInputText extends UIShape {
   @ObservableComponent.field
   paddingLeft: number = 0
 
-  @OnUUIDEvent.uuidEvent
   onTextSubmit: OnTextSubmit | null = null
 
-  @OnUUIDEvent.uuidEvent
   onChanged: OnChanged | null = null
 
+  // @internal
+  @OnUUIDEvent.uuidEvent
+  protected readonly onTextChanged: OnChanged
+
   @OnUUIDEvent.uuidEvent
   onFocus: OnFocus | null = null
 
   @OnUUIDEvent.uuidEvent
   onBlur: OnBlur | null = null
+
+  constructor(parent: UIShape | null) {
+    super(parent)
+    this.onTextChanged = new OnChanged((e) => {
+      const { value, isSubmit } = e.value
+
+      const isDirty = this.dirty
+      this.value = value
+      this.dirty = isDirty
+
+      if (isSubmit && this.onTextSubmit) {
+        this.onTextSubmit.callback({ text: value } as IEvents['onTextSubmit'])
```

Comment:
instead of using the casting `as`, its a good practice to declare the const with the strict types.

```ts
const value: IEvents['onTextSubmit'] = { text: value }
```
This way you cant add extra or invalid props to the event